### PR TITLE
feat: data aggregator

### PR DESCRIPTION
**Pull request by giovanecosta**

### CODE REVIEW

File: domain/network/network.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
+from dataclasses import asdict, dataclass
+from typing import List
+from datetime import datetime, timedelta
+from dacite import from_dict
+
+from domain.network.node import NodeState, Peer, Node
+
+
+def uptime_to_up_since(uptime: float) -> int:
```
Comment (msbrogli) [inline domain/network/network.py:9]: Missing docstring.

File: domain/network/network.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+    def add_connected_to(self, node_id: str) -> None:
+        self.connected_to.append(node_id)
+
+    def remove_connected_to(self, node_id: str) -> None:
+        self.connected_to.remove(node_id)
+
+
+@dataclass
+class AggregatedNode:
```
Comment (msbrogli) [inline domain/network/network.py:60]: What's the different between an AggregatePeer and an AggregateNode? Maybe add docstrings to both of them explaining in which case they are used and what for.

File: domain/network/network.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    def to_dict(self) -> dict:
+        """ Convert a AggregatedPeer instance into dict
+
+        :return: The dict representation of the AggregatedPeer
+        :rtype: dict
+        """
+        return asdict(self)
+
+    def add_connected_to(self, node_id: str) -> None:
+        self.connected_to.append(node_id)
```
Comment (msbrogli) [inline domain/network/network.py:53]: Maybe use a set? This can get duplicated entries.

File: text/0002-data-aggregator.md [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+We could do the aggregation on client side. However, more requests must be made every second, or always return everyhthing.
+
+The disadvantage of this approach is that we must keep polling. 
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- Nodes are still manually setted by us. It would be good to discover new http responding nodes automatically as they join the network.
```
Comment (msbrogli) [inline text/0002-data-aggregator.md:67]: Maybe it should get the list from an S3 location? Or a redis key? So we can build a separate script to update the list. @luislhl is working in a full-node discovery for other services. Maybe you can use the same solution with same flags to indicate which full-nodes will be used by each service.

File: text/0002-data-aggregator.md [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+WebSocket. Everytime `Network` have new content and is saved, a handler can be triggered to broadcast new data to connected clients.
```
Comment (msbrogli) [inline text/0002-data-aggregator.md:72]: I really don't see many advantages using websocket here since the uptime is changing all the time. And the API is so fast that it doesn't matter the hit rate.

File: text/0002-data-aggregator.md [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+[guide-level-explanation]: #guide-level-explanation
+
+Data aggregator will get data collected by `data-collector`, aggregate them into a Network dataclass, compare with saved data and save it if different.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Three new dataclasses will be created: `AggregatedNode`, `AggregatedPeer` and `Network`.
+
+- `AggregatedNode` and `AggregatedPeer` will be a bit smaller version of `Node` and `Peer` respectively, having `up_since` as the main change.
```
Comment (msbrogli) [inline text/0002-data-aggregator.md:27]: Nice. Add a description of each field and how they are aggregated, please.

Comment (msbrogli) [inline text/0002-data-aggregator.md:27]: I'm still trying to figure out why we need an `AggregatedNode` (and how it differs from an `AggregatedPeer`). Can you help me here, please?

File: text/0002-data-aggregator.md [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+We could do the aggregation on client side. However, more requests must be made every second, or always return everyhthing.
+
+The disadvantage of this approach is that we must keep polling. 
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- Nodes are still manually setted by us. It would be good to discover new http responding nodes automatically as they join the network.
+
+# Future possibilities
```
Comment (msbrogli) [inline text/0002-data-aggregator.md:69]: I think we should add that we'd like to show the peers in a world map. This is the coolest part of this aggregation in my point of view.

File: usecases/aggregate_node_data.py [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+        self.node_gateway = node_gateway or NodeGateway()
+
+    def aggregate(self) -> Union[bool, None]:
+        new_network = self.node_gateway.aggregate_network()
+        old_network = self.node_gateway.get_network()
+
+        networks_diff = DeepDiff(new_network, old_network)
+
+        if networks_diff:
+            return self.node_gateway.save_network(new_network)
```
Comment (msbrogli) [inline usecases/aggregate_node_data.py:19]: I'd say we would enter here "all the time". Should we store a hit counter to see if this diff is really necessary?

File: usecases/aggregate_node_data.py [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+from deepdiff import DeepDiff
+from gateways.node_gateway import NodeGateway
+
+
+class AggregateNodeData:
+
+    def __init__(self, node_gateway: Union[NodeGateway, None] = None) -> None:
+        self.node_gateway = node_gateway or NodeGateway()
+
+    def aggregate(self) -> Union[bool, None]:
```
Comment (msbrogli) [inline usecases/aggregate_node_data.py:12]: Missing docstring. What's the difference from returning a bool or a None?

File: usecases/aggregate_node_data.py [hunk 10]
```diff
@@ -0,0 +1,21 @@
+from typing import Union
+
+from deepdiff import DeepDiff
+from gateways.node_gateway import NodeGateway
+
+
+class AggregateNodeData:
```
Comment (msbrogli) [inline usecases/aggregate_node_data.py:7]: Missing docstring.

**Review by msbrogli — CHANGES_REQUESTED** [review]

**Comment by msbrogli** [comment]
I haven't reviewed all the files. I make a few comments in specific files but the most important comments are in the design.

File: domain/network/network.py [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+            id=peer.id,
+            app_version=peer.app_version,
+            up_since=uptime_to_up_since(peer.uptime),
+            address=peer.address,
+            state=peer.state,
+            last_message=int(peer.last_message),
+            latest_timestamp=peer.latest_timestamp,
+            sync_timestamp=peer.sync_timestamp,
+            warning_flags=peer.warning_flags,
+            entrypoints=[],  # TODO: add when Peer entrypoints is fixed HathorNetwork/hathor-explorer-service#28
```
Comment (pedroferreira1) [inline domain/network/network.py:40]: I didn't see this fix in the week tasks. I believe the entrypoints should be fixed before we set data aggregator as completed.

Comment (giovanecosta) [reply domain/network/network.py:40]: Good point, it's done now: https://github.com/HathorNetwork/hathor-explorer-service/pull/43

File: usecases/aggregate_node_data.py (hunk 9 above)
Comment (giovanecosta) [reply usecases/aggregate_node_data.py:12]: `StrictRedis set` has bool or None as returning type. It's just to match that. We can probably abstract that to only bool. I just didn't think about it.

File: text/0002-data-aggregator.md (hunk 4 above)
Comment (giovanecosta) [reply text/0002-data-aggregator.md:67]: Today we are using env vars. I like a lot this "full-node discovery" thing.

File: text/0002-data-aggregator.md (hunk 5 above)
Comment (giovanecosta) [reply text/0002-data-aggregator.md:72]: I'm proposing we use `up_since` now exactly to avoid this issue and optimize lambda usage and server-client communication

File: text/0002-data-aggregator.md (hunk 7 above)
Comment (giovanecosta) [reply text/0002-data-aggregator.md:69]: I'll add that

File: text/0002-data-aggregator.md (hunk 6 above)
Comment (giovanecosta) [reply text/0002-data-aggregator.md:27]: Aggregated Nodes and Peers has less properties and use `up_since` instead of `uptime`.

@pedroferreira1 Node is what we can connect and read the status. Peer is what we only know that exists because it is connected to a Node.

File: usecases/aggregate_node_data.py (hunk 8 above)
Comment (giovanecosta) [reply usecases/aggregate_node_data.py:19]: I don't think so, as we introducing `up_since` instead of `uptime` for comparison

File: text/0002-data-aggregator.md [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+
+Data aggregator will get data collected by `data-collector`, aggregate them into a Network dataclass, compare with saved data and save it if different.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Three new dataclasses will be created: `AggregatedNode`, `AggregatedPeer` and `Network`.
+
+- `AggregatedNode` and `AggregatedPeer` will be a bit smaller version of `Node` and `Peer` respectively, having `up_since` as the main change.
+  As `uptime` is always changing as time goes by, `up_since` remains the same.
```
[the remark this answers is not in the document]
Comment (giovanecosta) [reply text/0002-data-aggregator.md:28]: Oh, sure, that's why we'll use `up_since` so we can check the diff and the reconnection will be detected naturally.

File: text/0002-data-aggregator.md (hunk 6 above)
Comment (giovanecosta) [reply text/0002-data-aggregator.md:27]: Added

File: domain/network/network.py [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+    up_since_date = datetime.now() - timedelta(seconds=uptime)
+    return int(datetime.timestamp(up_since_date))
+
+
+@dataclass
+class AggregatedPeer:
+    """Aggregated data of a Peer retrieved by requesting a Node
+
+    :param id: id of peer
+    :type id: str
```
Comment (msbrogli) [inline domain/network/network.py:28]: I don't think we need the `type` anymore since we are using type annotations in the code. It will probably get outdated in the future.

File: domain/network/network.py [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+    :param entrypoints: list entrypoints, if any
+    :type entrypoints: List[str]
+
+    :param connected_to: list of ids of nodes that peer is conected to
+    :type connected_to: Set[str]
+
+    """
+    id: str
+    app_version: str
+    up_since: int
```
Comment (msbrogli) [inline domain/network/network.py:63]: Why store `up_since` instead of `uptime`?

File: tests/unit/usecases/test_collect_nodes_statuses.py [hunk 15]
```diff
@@ -20,6 +22,7 @@ def node_gateway(self):
         return MagicMock()
 
     @patch('usecases.collect_nodes_statuses.HATHOR_NODES', ['a', 'b'])
+    @freeze_time(datetime.now())
```
Comment (msbrogli) [inline tests/unit/usecases/test_collect_nodes_statuses.py:25]: Why do you need to freeze the time? I mean, you are not using a specific point in time in the tests.

File: usecases/aggregate_node_data.py (hunk 8 above)
Comment (msbrogli) [reply usecases/aggregate_node_data.py:20]: I got it. But only if we collect data twice or more in the same second, right? Otherwise, even the `up_since` will be different. Maybe we should create a comparison method that receives a tolerance parameter in seconds (e.g. `compareNodes(node1, node2, tol=60)`). What do you think?

**Review by msbrogli — CHANGES_REQUESTED** [review]

File: domain/network/network.py (hunk 2 above)
Comment (giovanecosta) [reply domain/network/network.py:115]: `Node` is what we can request, `Peer` is what we only have data through the `Node`. Aggregated versions follows the same concept. I've improved docstring making it clearer

File: domain/network/network.py (hunk 3 above)
Comment (giovanecosta) [reply domain/network/network.py]: done. thanks!

File: usecases/aggregate_node_data.py (hunk 8 above)
Comment (giovanecosta) [reply usecases/aggregate_node_data.py:20]: `up_since` is supposed to be fixed for a running node. It's the timestamp when it started to run.
As we calculate it in runtime using `uptime`,  two instances of `AggregatedNode` for the same node can be different for milliseconds. That's why `up_since` is also `int` instead of `float`.
We still have the edge case where this milliseconds difference also change the integer part of the second but i don't think it's a problem.

File: tests/unit/usecases/test_collect_nodes_statuses.py (hunk 15 above)
Comment (giovanecosta) [reply tests/unit/usecases/test_collect_nodes_statuses.py:25]: As we calculate `up_since` in runtime, it can change a second for two different instances of the same Node created in two different points of the code. It's an edge case that can bring intermittence to tests. So freeze the time will avoid it.

File: domain/network/network.py (hunk 14 above)
Comment (giovanecosta) [reply domain/network/network.py:63]: `uptime` keep changing every second. `up_since` is the timestamp of when the node started (`now - uptime`) so it wont change for the same running node.

File: domain/network/network.py (hunk 13 above)
Comment (giovanecosta) [reply domain/network/network.py:28]: Should we remove it? I don't have an opinion about it yet. As we're not generating any doc and the IDE is pretty much fine with the type hints, i think it doesn't matter to me, we can remove or keep it.

File: text/0002-data-aggregator.md (hunk 7 above)
Comment (giovanecosta) [reply text/0002-data-aggregator.md:69]: Done
